codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -600,7 +604,27 @@ protected void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Htt
      */
     protected void onStreamError(ChannelHandlerContext ctx, @SuppressWarnings("unused") Throwable cause,
                                  StreamException http2Ex) {
-        resetStream(ctx, http2Ex.streamId(), http2Ex.error().code(), ctx.newPromise());
+
+        int streamId = http2Ex.streamId();
+
+        //if this is caused by reading headers that are too large, send a header with status 431
+        if (http2Ex instanceof Http2Exception.DecodeHeaderListSizeException && connection().isServer()) {
+            final Http2Stream stream = connection().stream(streamId);
+
+            // NOTE We have to check to make sure that a stream exists before we send our reply.
+            // We likely always create the stream below as the stream isn't created until the
+            // header block is completely processed.
+            if (stream == null) {
+                try {
+                    encoder.connection().remote().createStream(streamId, true);
+                } catch (Http2Exception e) {
+                    resetUnknownStream(ctx, streamId, http2Ex.error().code(), ctx.newPromise());
+                }
+            }
+            frameWriter().writeHeaders(ctx, streamId, headersTooLarge, 0, true, ctx.newPromise());
```

Comment:
Could we flush here and then still do a `resetStream` so that we properly clean up this stream?